docs/MockFunctionAPI.md

```diff
@@ -130,7 +130,7 @@ Beware that `mockFn.mockClear()` will replace `mockFn.mock`, not just reset the
 
 Does everything that [`mockFn.mockClear()`](#mockfnmockclear) does, and also removes any mocked return values or implementations.
 
-This is useful when you want to completely reset a _mock_ back to its initial state.
+This is useful when you want to completely reset a _mock_ back to its initial state. (Note that resetting a _spy_ will result in a function with no return value).
```

Comment:
should it be re-added to older versions of the docs as well?

Reply:
Maybe. I'd reconsider the wording though, I think. The behaviour (pre 29.4.0 / after this PR) is to replace all spy mock implementations with empty functions - that is, it's not just about the return value.

```ts
test('mocks, spies, resetAllMocks', () => {
  jest.spyOn(console, 'log');
  jest.spyOn(globalThis, 'setImmediate').mockImplementation(cb => cb());

  const mockCb = jest.fn();
  setImmediate(mockCb); // Calls mockCb synchronously
  expect(mockCb).toHaveBeenCalled();

  // spyOn wraps original implementation, foo printed to console
  console.log('foo');
  expect(console.log).toHaveBeenCalledWith('foo');

  jest.resetAllMocks();

  // Replaces default spy implementation with a stub - nothing is printed.
  console.log('bar');
  expect(console.log.mock.calls).toEqual([['bar']]);

  setImmediate(mockCb); // Never calls mockCb
  expect(mockCb).not.toHaveBeenCalled();
});
```

This is in fact what the paragraph above the one in question already says it should do, and IMO the confusion only arises from "Back to its initial state", which leads to people expecting it to do what `mockRestore` does (and possibly led to https://github.com/jestjs/jest/issues/13808). 

Instead of:

> Does everything that [`mockFn.mockClear()`](#mockfnmockclear) does, and also removes any mocked return values or implementations.
> 
> This is useful when you want to completely reset a _mock_ back to its initial state.

How about:

> Does everything that [`mockFn.mockClear()`](#mockfnmockclear) does, and also replaces the mock implementation with an empty function, returning `undefined`.

Reply:
yeah good point. I think your suggestion sounds good 👍

Reply:
Updated and rebased 👍